compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -409,7 +409,40 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         let deref_ty = match *ty.kind() {
```

Comment:
Maybe have an assert after this entire thing that `deref_ty` is one of several allowed types (on which we may have to do this fallback). At least that way it couldn't be a reference to a reference or something silly like that.